### PR TITLE
fix: order by subquery planning

### LINKED ISSUE #16110: Incorrect Column Reference in ORDER BY Subquery Plan in vtgate
# Issue Description
## Problem
When performing an ORDER BY on a subquery, the vtgate planner generates a plan that does not work.

## Example
For the following query:

```sql
SELECT (SELECT col FROM user LIMIT 1) AS a 
FROM user 
JOIN user_extra 
ORDER BY a;
```

The planner produces a plan containing this snippet:

```json
{
  "OperatorType": "UncorrelatedSubquery",
  "Variant": "PulloutValue",
  "PulloutVars": [
    "__sq1"
  ],
  "Inputs": [
    {
      "InputName": "SubQuery",
      "OperatorType": "Limit",
      "Count": "1",
      "Inputs": [
        {
          "OperatorType": "Route",
          "Variant": "Scatter",
          "Keyspace": {
            "Name": "user",
            "Sharded": true
          },
          "FieldQuery": "select col from `user` where 1 != 1",
          "Query": "select col from `user` limit 1",
          "Table": "`user`"
        }
      ]
    },
    {
      "InputName": "Outer",
      "OperatorType": "Route",
      "Variant": "Scatter",
      "Keyspace": {
        "Name": "user",
        "Sharded": true
      },
      "FieldQuery": "select :__sq1 as __sq1, weight_string(:__sq1) from `user` where 1 != 1",
      "OrderBy": "(0|1) ASC",
      "Query": "select :__sq1 as __sq1, weight_string(:__sq1) from `user` order by __sq1 asc",
      "Table": "`user`"
    }
  ]
}
```

## Issue
In the query sent to the outer side of the subquery, the plan uses the column name __sq1 instead of the argument :__sq1, which would be the correct plan.

## Result
This discrepancy results in a query that fails on MySQL, as no such column (`__sq1`) can be found.

### PR DESCRIPTION
**Pull request by harshit-gangal**
## Description
This PR fixes an issue in the vtgate planner where an incorrect column reference is used in ORDER BY subquery plans.

### Fix
The planner now correctly uses arguments instead of the column names for subquerie placeholders, ensuring the query works as expected on MySQL.

## Related Issue(s)
Fixes #16110
The regression was introduced in #15039.

## Checklist
-   [x] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [x] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [x] Tests were added or are not required
-   [x] Did the new or modified tests pass consistently locally and on CI?
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/planbuilder/operators/subquery_planning.go [hunk 1]
```diff
@@ -132,7 +157,7 @@ func rewriteMergedSubqueryExpr(ctx *plancontext.PlanningContext, se SubQueryExpr
 							return true
 						}
 					case *sqlparser.Argument:
-						if expr.Name != sq.ArgName {
+						if expr.Name != sq. /**/ ArgName {
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/operators/subquery_planning.go:160]: nit: comment at wrong place

File: go/vt/vtgate/planbuilder/testdata/select_cases.json [hunk 2]
```diff
@@ -2278,70 +2278,60 @@
       "QueryType": "SELECT",
       "Original": "select (select col from user limit 1) as a from user join user_extra order by a",
       "Instructions": {
-        "OperatorType": "SimpleProjection",
```
Comment (systay) [inline go/vt/vtgate/planbuilder/testdata/select_cases.json:2281]: This plan was not working before these changes. Here we are ordering by the result of a subquery

File: go/vt/vtgate/planbuilder/operators/subquery.go [hunk 3]
```diff
@@ -293,3 +296,31 @@ func (sq *SubQuery) mapExpr(f func(expr sqlparser.Expr) sqlparser.Expr) {
...
+	if !ok {
+		return sq, NoRewrite
+	}
+	if other.ArgName == sq.ArgName {
+		sq.Outer = other.Outer
+		return sq, Rewrote("removed duplicate subquery")
+	}
+
+	return sq, NoRewrite
+}
```
Comment (harshit-gangal) [inline go/vt/vtgate/planbuilder/operators/subquery.go:326]: I do not think we ever encounter this case. This can be removed.

**Review by frouioui — COMMENTED** [review]
It would be really great to have the `.test` file directly in the repo or have it easily testable in the PR.

**Comment by frouioui** [comment]
Otherwise the vitess-tester `subqueries.txt` seem to be passing fine on my end besides the evalengine type issue and the `VT12001` error.

**Review by harshit-gangal — COMMENTED** [review]
need more e2e and plan tests

**Comment by systay** [comment]
> It would be really great to have the `.test` file directly in the repo or have it easily testable in the PR.

Yeah, but it also feels like we are complecting things that don't need to go together.

File: go/vt/vtgate/planbuilder/operators/aggregator.go [hunk 4]
```diff
@@ -379,6 +379,21 @@ func (a *Aggregator) planOffsets(ctx *plancontext.PlanningContext) Operator {
 	return nil
 }
 
+func (aggr Aggr) setPushColumn(exprs sqlparser.Exprs) {
+	if aggr.Func == nil {
+		if len(exprs) > 1 {
+			panic(vterrors.VT13001(fmt.Sprintf("unexpected number of expression in an random aggregation: %s", sqlparser.String(exprs))))
```
[the remark this answers is not in the document]
Comment (systay) [reply go/vt/vtgate/planbuilder/operators/aggregator.go:385]: no, this is a bug. we got the arguments, and rewrote them, and that should not change the number of arguments

**Review by systay — APPROVED** [review]
Accepting my own changes with @harshit-gangal's blessing

**Review by frouioui — APPROVED** [review]

**Comment by frouioui** [comment]
Let's backport this PR and https://github.com/vitessio/vitess/pull/15039 to `release-18.0`. There is a stale backport PR that got closed (https://github.com/vitessio/vitess/pull/15068). Let's only do one backport with both commits.

**Comment by harshit-gangal** [comment]
> Let's backport this PR and #15039 to `release-18.0`. There is a stale backport PR that got closed (#15068). Let's only do one backport with both commits.

This PR as-is cannot be ported to v18. v18 have different way of handling subqueries

**Comment by frouioui** [comment]
We should change the `Skip` in the end-to-end test to allow the test to run.

**Pull request merged**